File: dp/c/pipes_and_filters.c
// Pipes and Filters Pattern in C

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef void (*FilterFn)(int* data, int* size);

void filter_remove_negative(int* data, int* size) {
    int j = 0;
    for (int i = 0; i < *size; i++)
        if (data[i] >= 0) data[j++] = data[i];
    *size = j;
    printf("[Filter] Remove negatives -> %d items\n", *size);
}

void filter_double(int* data, int* size) {
    for (int i = 0; i < *size; i++) data[i] *= 2;
    printf("[Filter] Double values\n");
}

void filter_keep_under_100(int* data, int* size) {
    int j = 0;
    for (int i = 0; i < *size; i++)
        if (data[i] < 100) data[j++] = data[i];
    *size = j;
    printf("[Filter] Keep < 100 -> %d items\n", *size);
}

void filter_sort(int* data, int* size) {
    for (int i = 0; i < *size - 1; i++)
        for (int j = i + 1; j < *size; j++)
            if (data[i] > data[j]) { int t = data[i]; data[i] = data[j]; data[j] = t; }
    printf("[Filter] Sort\n");
}

typedef struct {
    FilterFn filters[16];
    int count;
} Pipeline;

void pipeline_add(Pipeline* p, FilterFn f) { p->filters[p->count++] = f; }

void pipeline_run(Pipeline* p, int* data, int* size) {
    for (int i = 0; i < p->count; i++)
        p->filters[i](data, size);
}

int main(void) {
    int data[] = {42, -5, 10, 80, 3, -1, 55, 25};
    int size = 8;

    Pipeline p = {.count = 0};
    pipeline_add(&p, filter_remove_negative);
    pipeline_add(&p, filter_double);
    pipeline_add(&p, filter_keep_under_100);
    pipeline_add(&p, filter_sort);

    pipeline_run(&p, data, &size);

    printf("Result:");
    for (int i = 0; i < size; i++) printf(" %d", data[i]);
    printf("\n");
    return 0;
}

File: dp/rust/src/pipes_and_filters.rs
// Pipes and Filters Pattern in Rust
// Chain processing steps using iterators / closures

type Filter<T> = Box<dyn Fn(Vec<T>) -> Vec<T>>;

struct Pipeline<T> {
    filters: Vec<Filter<T>>,
}

impl<T: 'static> Pipeline<T> {
    fn new() -> Self {
        Pipeline { filters: Vec::new() }
    }

    fn add_filter<F>(mut self, f: F) -> Self
    where
        F: Fn(Vec<T>) -> Vec<T> + 'static,
    {
        self.filters.push(Box::new(f));
        self
    }

    fn execute(self, input: Vec<T>) -> Vec<T> {
        self.filters.into_iter().fold(input, |data, filter| filter(data))
    }
}

// Text processing pipeline
fn main() {
    // Numeric pipeline
    println!("=== Numeric Pipeline ===");
    let result = Pipeline::new()
        .add_filter(|nums: Vec<i32>| {
            println!("[Filter] Remove negatives");
            nums.into_iter().filter(|&n| n >= 0).collect()
        })
        .add_filter(|nums| {
            println!("[Filter] Double values");
            nums.into_iter().map(|n| n * 2).collect()
        })
        .add_filter(|nums| {
            println!("[Filter] Keep < 100");
            nums.into_iter().filter(|&n| n < 100).collect()
        })
        .add_filter(|mut nums| {
            println!("[Filter] Sort");
            nums.sort();
            nums
        })
        .execute(vec![42, -5, 10, 80, 3, -1, 55, 25]);

    println!("Result: {:?}", result);

    // String pipeline
    println!("\n=== Text Pipeline ===");
    let texts = vec![
        "  Hello World  ".to_string(),
        "RUST programming".to_string(),
        "  ".to_string(),
        "pipes AND filters".to_string(),
    ];

    let processed = Pipeline::new()
        .add_filter(|v: Vec<String>| {
            println!("[Filter] Trim");
            v.into_iter().map(|s| s.trim().to_string()).collect()
        })
        .add_filter(|v| {
            println!("[Filter] Remove empty");
            v.into_iter().filter(|s| !s.is_empty()).collect()
        })
        .add_filter(|v| {
            println!("[Filter] Lowercase");
            v.into_iter().map(|s| s.to_lowercase()).collect()
        })
        .execute(texts);

    println!("Result: {:?}", processed);
}
